### PR TITLE
Merge PeekStream and BufferedInStream

### LINKED ISSUE #4501: provide a peeking API to std.io.BufferedReader
The point of BufferedInStream is to call `read()` on the underlying stream in bursts, rather than separate calls. E.g. to make 1 syscall to `read()` reading 999 bytes, rather than 999 syscalls to `read()` reading 1 byte each.

The point of PeekStream is to be able to "put back" bytes, so that the next call to `read()` will get them again. In order to accomplish this, PeekStream needs a buffer. That makes PeekStream also a BufferedInStream.

All we need to unify these two APIs is for BufferedInStream to have a parameter called `reserved_buffer_space` or something like this, and then it's guaranteed to be able to "put back" this many elements.

Related:
* #19298

### PR DESCRIPTION
**Pull request by daurnimator**
Child of #4505
Closes #4501

### CODE REVIEW

File: lib/std/io/buffered_in_stream.zig [hunk 1]
```diff
@@ -3,16 +3,46 @@ const io = std.io;
...
+                        .unbuffered_in_stream = unbuffered_in_stream,
+                        .fifo = FifoType.init(buf),
+                    };
+                }
+            },
+            .Dynamic => struct {
+                pub fn init(unbuffered_in_stream: InStreamType, allocator: *mem.Allocator) Self {
+                    return .{
+                        .unbuffered_in_stream = unbuffered_in_stream,
+                        .fifo = FifoType.init(allocator),
```
Comment (joachimschmidt557) [inline lib/std/io/buffered_in_stream.zig:41]: This dynamic FIFO needs to be deinitialised at some time. Would it make sense to add a special `BufferedInStream.deinit` function for this?

**Comment by joachimschmidt557** [comment]
I'm very unsure about this, but: Isn't the behaviour of `PeekStream.read` different from `BufferedInStream.read`? A `PeekStream.read` never fills the FIFO, only `PeekStream.putBack` can do that. But a `BufferedInStream` does automatically add to the FIFO while reading to speed things up. 

Couldn't that create a problem when I try to `putBack` something and it doesn't work with the merged API because `read` already filled the FIFO to its limit e.g. when I have a lookahead of exactly 1 byte?

**Comment by daurnimator** [comment]
> I'm very unsure about this, but: Isn't the behaviour of `PeekStream.read` different from `BufferedInStream.read`? A `PeekStream.read` never fills the FIFO, only `PeekStream.putBack` can do that. But a `BufferedInStream` does automatically add to the FIFO while reading to speed things up.

Yes.

> Couldn't that create a problem when I try to putBack something and it doesn't work with the merged API because read already filled the FIFO to its limit e.g. when I have a lookahead of exactly 1 byte?

Ah, prior experience has made me take something for granted here: in the traditional libc/POSIX `ungetc` api, you are only allowed to put back what you last read: a subsequent `read` makes it undefined if you can fit things in or not. I did one-better than the libc API by at least allowing `putBack` to return an error if it doesn't fit: additionally, if you use a `Dynamic` buffer, then it will attempt to grow.

**Comment by andrewrk** [comment]
Feel free to re-open with feedback addressed & rebased. It also sounds like it may be worth discussing on #4501 before implementing, since this takes the accepted proposal in a different direction.

**Pull request closed without merging**